src/main/java/bio/terra/workspace/service/workspace/flight/CreateWorkspaceAuthzStep.java

```diff
@@ -52,9 +55,14 @@ public StepResult undoStep(FlightContext flightContext) {
     FlightMap inputMap = flightContext.getInputParameters();
     // Only delete the Sam resource if we actually created it in the do step.
     FlightMap workingMap = flightContext.getWorkingMap();
-    if (workingMap.get(AUTHZ_COMPLETED_KEY, Boolean.class)) {
-      UUID workspaceID = inputMap.get(WorkspaceFlightMapKeys.WORKSPACE_ID, UUID.class);
+    UUID workspaceID = inputMap.get(WorkspaceFlightMapKeys.WORKSPACE_ID, UUID.class);
+    if (workingMap.get(AUTHZ_COMPLETED_KEY, Boolean.class) != null) {
```

Comment:
Where we set this value now? And do we still need this?

Reply:
This value still gets set after creating a workspace resource in Sam, and we still need it to confirm that the doStep completed and created that resource.

Reply:
ah, I see. Sorry, I just saw you remove `workingMap.put(AUTHZ_COMPLETED_KEY, false);` and thought this is not needed. Feels little bit wired that we are not actually using the `Boolean` type, this line works with any other types, all we need is to put something in working map. But I think it's fine.

Also, we still have the same problem as I discussed in my doc(working map is step boundary,)? 
Not famialr with the permormance requirement/use cases, but other options I can come up with if we want to get rid of working map here:
1. Check existances before calling delete
2. Call delete and catch 404 exception.

Reply:
Yeah, it feels a little silly to only be using a "present/absent" flag, but I guess Boolean is the right type for that?

Agreed, this still has the issue of orphaning resources when there's a failure between the actual creation and marking the value in the working map - at least we have a warning for it now. I think both of your options run the risk of deleting resources that weren't created in this flight: since WorkspaceIDs are passed in externally via request, multiple flights might be running for the same ID. 

Without some kind of locking mechanism, I don't think we can guarantee that flights will both 1. undo everything they did and 2. won't undo work they didn't do. Since #2 is much worse than #1, that's the one this workingMap key prevents.

Reply:
drive-by locking mechanism comment: the workspaceId can be used as the jobId in job service for creating workspaces. Then, we can ensure that at most one create workspace request is running for a given id. In general, we want users to specify unique id for long running operations and de-dupe logical requests to the same operation. It's a little awkward atm to check the JobService for (is this jobId new, or does the user have access to it, or is the user not authorized.)

Reply:
How does stairway handle multiple flights with the same ID - does it throw an exception for the second one, or just queue it and run it after the first one finishes? 

That does seem useful here but if we expect to use that to support concurrency elsewhere, it feels like what we really want is locking.

Reply:
Flight ids need to be unique until they're manually deleted from Stairway. atm, old flights accumulated in stairway db. It will throw an exception since it violates the unique id constraint of its db. We can write a unittest to determine exactly what it looks like atm.

Reply:
I think Yonghao's right that we should either check existence before calling delete or handle 404 here. We should ensure that no more than one flight is trying to create a workspace with the same id simultaneously somewhere else, whether that's locking or using the workspaceId as the job/flight id.

I think we will want locking for workspaces, but I'm not sure how much that helps us here - we'd have to lock a workspace before it's created, or create the workspace in a-not-ready-to-get stage. Using JobService/Stairway unique guarantees seems easier, and we will also want that functionality for other operations, but open to other ideas.

Reply:
The old flights accumulating concerns me. I don't like the idea that WM won't ever be able to create a particular workspace because e.g. Sam was temporarily down when we tried to make the workspace. That's especially bad because we're taking WorkspaceIDs in from requests, since we can't just internally try again with a new ID and clients would now need to know that behavior.

Reply:
Good point. We can include a operation id in the create workspace request so that a client that's trying to do the same logical operation does the same thing, but at a certain point we do need to worry about different logical requests deciding on the same workspace id, or say that's random enough to not worry about.

Reply:
I'm not really worried about different logical requests picking the same workspace ID - there aren't duplicate workspace IDs in existing Terra, and if someone tries to create one a 400 response makes sense. In general I expect WM createWorkspace calls to come from a single place (either Rawls or folder manager, depending) so I'm also less worried about systems that don't know they're making simultaneous calls.

Passing in an operation id and (I'm assuming) using that as the flight ID does solve this problem for WM, but it's really bumping it up a layer by expecting clients to manage their logical operations. That might be the right call, but it means a more complicated endpoint and changes to client behavior. I think locking is the right behavior here, even if it means locking on a not-totally-created workspace.

Reply:
While we'll also need locking, I will argue that operation ids are actually simpler for a client that's trying to handle failure. And we can always make the operation id optional and generate it if it's not specified.
If a client of WM doesn't generate it's only operation id, consider what happens when there's a network failure in the call to create the workspace. Now the client has to resend the request to create the workspace. What should they do when they receive a 400 for the second request? Do they try to track that they already send a request to create a workspace and treat 400 as ok in that special case? Do they error out and orphan the workspace? Do they infer that there operation has already begun and try to poll as if they received a 200?
If there is a generated operation id, WM can 200 duplicate requests, deduping the logical requests. The client behavior is always the same no matter which request succeeded in creating the operation in WM.
If the client is a stairway user, e.g. folder manager, generating and persisting the operation id before starting the operation is the familiar pattern, just as we're doing for data reference id generation. In this case, the client is pre-generating the workspace id. We could generate an additional operation id, but the workspace id fulfills the purpose of uniquely identifying the logical operation.
It's true that this chain of operation id generation has to stop at some client, e.g. the user's browser or the client not generating an id at all, but we should be careful when we cutoff the ability to specify an operation id.

Reply:
That's a good point about idempotency, especially for clients using stairway. In that case, I'll update the whole request flow and probably bump the client version as well.